### PR TITLE
Bug fixing in ADMM + code optimization as discussed

### PR DESCRIPTION
**Pull request by adekusar-drl**
### Summary
- Fixed bugs in ADMM, added unit mores
- Code optimization as discussed (see below)
- Minor changes according to the latest discussion/review.

### Details and comments
- Fixed maximization support in ADMMOptimizer and a new test case added based on the example provided.
- Fixed some other issues in ADMMOptimizer
- Added reference to the paper on ADMM
- QUBO / Cts. Optimizer out of ADMMParams and as attributes of ADMMOptimizer
- ADMMParams is now an internal field in optimizer
- ADMMOptimizerResult is a class derived from OptimizationResult and a new state property is introduced
- Minor changes according to the latest discussion/review.

### CODE REVIEW

File: qiskit/optimization/algorithms/admm_optimizer.py [hunk 1]
```diff
@@ -156,37 +149,49 @@ def __init__(self,
         self.rho = rho_initial
 
 
+class ADMMOptimizerResult(OptimizationResult):
+    """ ADMMOptimizer Result."""
+
+    def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
```
Comment (Cryoris) [inline qiskit/optimization/algorithms/admm_optimizer.py:155]: Isn't the type of `x` a `np.ndarray` and of `fval` a float? Or does ADMM return different types sometimes?

**Comment by Cryoris** [comment]
Hi Anton, these changes look good, thanks! But before we merge could you change the base branch to `optimization-stack` (see [1])? Manoel did general updates on lint and doc errors and this way we can make sure that we'll get no merge-conflicts. 

[1]: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request

File: qiskit/optimization/algorithms/admm_optimizer.py (hunk 1 above)
Comment (adekusar-drl) [reply qiskit/optimization/algorithms/admm_optimizer.py:155]: You are right. But in terms of inheritance they are of type `Optional[Any]`. Changing the type in the constructor does not make much sense without overriding property methods from the base class.

**Comment by adekusar-drl** [comment]
Changed branch to optimization-stack, now there are conflicts.

**Comment by Cryoris** [comment]
Yes, that's what I thought, could you pull from `optimization-stack` and resolve them?

**Comment by adekusar-drl** [comment]
@Cryoris done.

**Comment by Cryoris** [comment]
Thank you for fixing it so quickly!

**Pull request merged**